httpcore5-reactive/src/main/java/org/apache/hc/core5/reactive/ReactiveResponseConsumer.java

```diff
@@ -0,0 +1,176 @@
+
+/**
+ * An {@link AsyncResponseConsumer} that publishes the response body through
+ * a {@link Publisher}, as defined by the Reactive Streams specification. The
+ * response is represented as a {@link Message} consisting of a {@link
+ * HttpResponse} representing the headers and a {@link Publisher} representing
+ * the response body as an asynchronous stream of {@link ByteBuffer} instances.
+ *
+ * @since 5.0
+ */
+@Contract(threading = ThreadingBehavior.SAFE)
+public final class ReactiveResponseConsumer implements AsyncResponseConsumer<Void> {
+
+    private final ReactiveDataConsumer reactiveDataConsumer = new ReactiveDataConsumer();
+    private final List<Header> trailers = Collections.synchronizedList(new ArrayList<Header>());
+    private final BasicFuture<Message<HttpResponse, Publisher<ByteBuffer>>> responseFuture;
+
+    private volatile BasicFuture<Void> responseCompletion;
+    private volatile HttpResponse informationResponse;
+    private volatile EntityDetails entityDetails;
+
+    /**
+     * Creates a {@code ReactiveResponseConsumer}.
+     */
+    public ReactiveResponseConsumer() {
+        this.responseFuture = new BasicFuture<>(null);
+    }
+
+    /**
+     * Creates a {@code ReactiveResponseConsumer} that will call back the supplied {@link FutureCallback} with a
+     * streamable response.
+     *
+     * @param responseCallback the callback to invoke when the response is available for consumption.
+     */
+    public ReactiveResponseConsumer(final FutureCallback<Message<HttpResponse, Publisher<ByteBuffer>>> responseCallback) {
+        this.responseFuture = new BasicFuture<>(Args.notNull(responseCallback, "responseCallback"));
+    }
+
+    public Future<Message<HttpResponse, Publisher<ByteBuffer>>> getResponseFuture() {
+        return responseFuture;
+    }
+
+    /**
+     * Returns the intermediate (1xx) HTTP response if one was received.
+     *
+     * @return the information response, or {@code null} if none.
+     */
+    public HttpResponse getInformationResponse() {
+        return informationResponse;
+    }
+
+    /**
+     * Returns the response entity details.
+     *
+     * @return the entity details, or {@code null} if none.
+     */
+    public EntityDetails getEntityDetails() {
+        return entityDetails;
+    }
+
+    /**
+     * Returns the trailers received at the end of the response.
+     *
+     * @return a non-null list of zero or more trailers.
+     */
+    public List<Header> getTrailers() {
+        return trailers;
+    }
+
+    @Override
+    public void consumeResponse(
+        final HttpResponse response,
+        final EntityDetails entityDetails,
+        final HttpContext httpContext,
+        final FutureCallback<Void> resultCallback
+    ) {
+        this.entityDetails = entityDetails;
+        this.responseCompletion = new BasicFuture<>(resultCallback);
+        this.responseFuture.completed(new Message<HttpResponse, Publisher<ByteBuffer>>(response, reactiveDataConsumer));
```

Comment:
@rschmitt Why not using a simple `Consumer` instead? Do we really need another `Future` here?

Reply:
I'm not sure what you mean by `Consumer`, because I thought this entire project was targeting JDK7. (RxJava provides a `Consumer<T>` interface, but the Reactive Streams API does not.) Either way, I would argue that we *do* need to return a `Future`, for the same reason that the `#execute` method returns a `Future`: callers need a single thing that they can block on (or decorate with callback logic) to wait for success, failure, or cancellation.

In the integration tests, there is a fairly consistent idiom:

```java
        final ReactiveResponseConsumer consumer = new ReactiveResponseConsumer();
        requester.execute(request, consumer, Timeout.ofSeconds(2), null);
        final Message<HttpResponse, Publisher<ByteBuffer>> response = consumer.getResponseFuture().get();
```

With a `Consumer`-based API, a caller might write:

```java
        final BlockingQueue<Message<HttpResponse, Publisher<ByteBuffer>>> streamingResponse =
            new ArrayBlockingQueue<>(1);
        final ReactiveResponseConsumer consumer =
            new ReactiveResponseConsumer(new Consumer<Message<HttpResponse, Publisher<ByteBuffer>>>() {
                @Override
                public void accept(final Message<HttpResponse, Publisher<ByteBuffer>> httpResponsePublisherMessage) {
                    streamingResponse.offer(httpResponsePublisherMessage);
                }
            });
        requester.execute(request, consumer, Timeout.ofSeconds(2), null);
        final Message<HttpResponse, Publisher<ByteBuffer>> response = streamingResponse.take();
```

However, the above code will block forever if an exception occurs and the response fails to come back. One way to fix this is by setting up a `CompletableFuture` or `BasicFuture` and then completing it from *both* callbacks:

```java
        final CompletableFuture<Message<HttpResponse, Publisher<ByteBuffer>>> streamingResponse =
            new CompletableFuture<>();
        final ReactiveResponseConsumer consumer =
            new ReactiveResponseConsumer(new Consumer<Message<HttpResponse, Publisher<ByteBuffer>>>() {
                @Override
                public void accept(final Message<HttpResponse, Publisher<ByteBuffer>> httpResponsePublisherMessage) {
                    streamingResponse.complete(httpResponsePublisherMessage);
                }
            });
        requester.execute(request, consumer, Timeout.ofSeconds(2), new FutureCallback<Void>() {
            @Override
            public void completed(Void result) {
                // TODO: Can this race with the other callback?
                streamingResponse.completeExceptionally(new RuntimeException("No stream returned"));
            }

            @Override
            public void failed(Exception ex) {
                streamingResponse.completeExceptionally(ex);
            }

            @Override
            public void cancelled() {
                streamingResponse.completeExceptionally(new CancellationException());
            }
        });
        final Message<HttpResponse, Publisher<ByteBuffer>> response = streamingResponse.join();
```

In short, constructing a correctly functioning `Future` for the streaming response is definitely something I want to do on the caller's behalf, rather than asking the caller to construct one manually from lower-level primitives.